### PR TITLE
Add CLI arguments processing

### PR DESCRIPTION
**Pull request by e2e4b6b7**

### CODE REVIEW

File: plugin/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    val json: String? by project
+
+    init {
+        jvmArgs = listOf(
+            "-Djava.awt.headless=true",
+            "--add-exports",
+            "java.base/jdk.internal.vm=ALL-UNNAMED",
+            "-Djdk.module.illegalAccess.silent=true"
+        )
+        maxHeapSize = "2g"
```
Comment (onewhl) [inline plugin/build.gradle.kts]: is it enough?

File: plugin/src/main/java/org/jetbrains/research/refactoringminer/RefactoringMiner.java [hunk 2]
```diff
@@ -1,19 +1,172 @@
...
+                case "gc":
+                    detectAtGitHubCommit(out, args.get(PATH), args.get(POSITION), Integer.parseInt(args.get(TIMEOUT)));
+                    break;
+                case "gp":
+                    detectAtGitHubPullRequest(out, args.get(PATH),
+                        Integer.parseInt(args.get(POSITION)), Integer.parseInt(args.get(TIMEOUT)));
+                    break;
+            }
+        } catch (IllegalArgumentException e) {
+            System.out.println("Call help to show usage.");
```
Comment (onewhl) [inline plugin/src/main/java/org/jetbrains/research/refactoringminer/RefactoringMiner.java:65]: if you want to log something, use a logger please

File: plugin/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,49 @@
+dependencies {
+    implementation(project(":refactoringminer"))
+}
+
+open class IOCliTask : org.jetbrains.intellij.tasks.RunIdeTask() {
+    // Name of operation to perform (h/help for usage)
+    val operation: String? by project
+
+    // Path to the local git directory or URL
+    val pathToGit: String? by project
```
Comment (onewhl) [inline plugin/build.gradle.kts]: why do we need it? or you mean a path to the git repository?

File: plugin/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+    // Commit, tag, branch or pull-request
+    val position: String? by project
```
Comment (onewhl) [inline plugin/build.gradle.kts]: startPosition?

File: plugin/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    // Commit or tag for bt and bc options
+    val endPosition: String? by project
+
+    val timeout: String? by project
+
+    // Path to json for output
+    val json: String? by project
```
Comment (onewhl) [inline plugin/build.gradle.kts]: maybe call it `output`?

File: plugin/src/main/java/org/jetbrains/research/refactoringminer/OutputRefactoringHandler.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
+package org.jetbrains.research.refactoringminer;
+
+import org.jetbrains.annotations.NotNull;
+import org.refactoringminer.api.Refactoring;
+import org.refactoringminer.api.RefactoringHandler;
+import org.refactoringminer.rm1.GitHistoryRefactoringMinerImpl;
+import java.io.IOException;
+import java.util.List;
+
+public class OutputRefactoringHandler extends RefactoringHandler {
```
Comment (onewhl) [inline plugin/src/main/java/org/jetbrains/research/refactoringminer/OutputRefactoringHandler.java:12]: what is the purpose of this class? i'm not sure about the choice of the name

File: plugin/src/main/java/org/jetbrains/research/refactoringminer/OutputRefactoringHandler.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        if (verbose) {
+            System.out.printf("Total count: [Commits: %d, Errors: %d, Refactorings: %d]%n",
+                commitsCount, errorCommitsCount, refactoringsCount);
+        }
+    }
+
+    @Override
+    public void handleException(String commit, Exception e) {
+        System.err.println("Error processing commit " + commit);
+        e.printStackTrace();
```
Comment (onewhl) [inline plugin/src/main/java/org/jetbrains/research/refactoringminer/OutputRefactoringHandler.java]: it's not good. please, use a logger to log some information, you can also add to the log some details about the excpetion

File: plugin/src/main/java/org/jetbrains/research/refactoringminer/RefactoringMiner.java [hunk 8]
```diff
@@ -1,19 +1,172 @@
...
         return "RefactoringMiner";
     }
 
     @Override
     public void main(@NotNull List<String> args) {
-        System.out.println("Empty main");
+        args = args.stream().map(Strings::emptyToNull).collect(Collectors.toList());
+        Path jsonPath = args.get(JSON) != null ? Path.of(args.get(JSON)) : null;
+        try (JsonOutput out = new JsonOutput(jsonPath)) {
+            switch (args.get(OPERATION)) {
```
Comment (onewhl) [inline plugin/src/main/java/org/jetbrains/research/refactoringminer/RefactoringMiner.java:37]: how does it match with the plugin's arguments?

**Review by onewhl — CHANGES_REQUESTED** [review]
thank you! please, take a look at the comments i left

File: plugin/build.gradle.kts (hunk 3 above)
Comment (e2e4b6b7) [reply plugin/build.gradle.kts]: Yes. Not good naming?

File: plugin/src/main/java/org/jetbrains/research/refactoringminer/OutputRefactoringHandler.java (hunk 6 above)
Comment (e2e4b6b7) [reply plugin/src/main/java/org/jetbrains/research/refactoringminer/OutputRefactoringHandler.java:12]: To handle finished commit or exceptions and print them

File: plugin/build.gradle.kts (hunk 3 above)
Comment (onewhl) [reply plugin/build.gradle.kts]: i thought it is a path to the user's git root. what about `"projectPath"`?

Comment (e2e4b6b7) [reply plugin/build.gradle.kts]: Ok

Comment (e2e4b6b7) [reply plugin/build.gradle.kts]: Gradle throw exception on this name. I think it is reserved. Suggest `gitProjectPath`

Comment (onewhl) [reply plugin/build.gradle.kts]: looks good to me

File: plugin/src/main/java/org/jetbrains/research/refactoringminer/OutputRefactoringHandler.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        this.repositoryURL = repositoryURL;
+    }
+
+    @Override
+    public void handle(String revisionId, List<Refactoring> refactorings) {
+        String revisionURL = GitHistoryRefactoringMinerImpl.extractCommitURL(repositoryURL, revisionId);
+        try {
+            output.commit(repositoryURL, revisionURL, revisionId, refactorings);
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (onewhl) [inline plugin/src/main/java/org/jetbrains/research/refactoringminer/OutputRefactoringHandler.java]: please, avoid `e.printStackTrace()` in code

File: plugin/src/main/java/org/jetbrains/research/refactoringminer/OutputRefactoringHandler.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+import org.jetbrains.annotations.NotNull;
+import org.refactoringminer.api.Refactoring;
+import org.refactoringminer.api.RefactoringHandler;
+import org.refactoringminer.rm1.GitHistoryRefactoringMinerImpl;
+import java.io.IOException;
+import java.util.List;
+
+public class OutputRefactoringHandler extends RefactoringHandler {
+    private final boolean verbose;
```
Comment (onewhl) [inline plugin/src/main/java/org/jetbrains/research/refactoringminer/OutputRefactoringHandler.java:15]: what is the purpose of this variable?

File: plugin/build.gradle.kts [hunk 11]
```diff
@@ -0,0 +1,49 @@
+dependencies {
+    implementation(project(":refactoringminer"))
+}
+
+open class IOCliTask : org.jetbrains.intellij.tasks.RunIdeTask() {
+    // Name of operation to perform (h/help for usage)
+    val operation: String? by project
```
Comment (onewhl) [inline plugin/build.gradle.kts]: maybe `command`?

File: plugin/src/main/java/org/jetbrains/research/refactoringminer/OutputRefactoringHandler.java (hunk 10 above)
Comment (e2e4b6b7) [reply plugin/src/main/java/org/jetbrains/research/refactoringminer/OutputRefactoringHandler.java:15]: Whether to print the commit statistics at the end of processing

**Review by onewhl — APPROVED** [review]
Looks good to me, thank you!

**Pull request merged**